src/librbd/image/OpenRequest.cc

```diff
@@ -523,6 +523,41 @@ Context *OpenRequest<I>::handle_set_snap(int *result) {
     return nullptr;
   }
 
+  // TODO SetSnapRequest should handle (re)-initializing image cache
+  return m_on_finish;
+}
+
+template <typename I>
+Context *OpenRequest<I>::send_init_image_cache(int *result) {
+  if (m_image_ctx->old_format || m_image_ctx->read_only ||
+      !m_image_ctx->persistent_cache_enabled) {
+    *result = 0;
+    return m_on_finish;
+  }
+
+  CephContext *cct = m_image_ctx->cct;
+  ldout(cct, 10) << this << " " << __func__ << dendl;
+
+  // TODO: hard-coded FileImageCache for prototype
+  m_image_ctx->image_cache = new cache::FileImageCache<ImageCtx>(*m_image_ctx);
```

Comment:
The read-only cache is being opened for images that are not read-only here. This logic should be moved from here to ```SetSnapRequest``` so that the image cache is opened and closed when a snapshot is set / unset (respectively). That will result in a cache being initialized for each parent image in the clone hierarchy (since this is for a read-only parent cache).